File: src/utils/cpp/evio_merge_files.cpp
//
// Copyright 2025, Jefferson Science Associates, LLC.
// Subject to the terms in the LICENSE file found in the top-level directory.
//
// EPSCI Group
// Thomas Jefferson National Accelerator Facility
// 12000, Jefferson Ave, Newport News, VA 23606
// (757)-269-7100

#include <string>
#include <cstdint>
#include <cstdlib>
#include <cstdio>
#include <chrono>
#include <memory>
#include <limits>
#include <unistd.h>
#include <random>
#include <iostream>

#include "eviocc.h"

using namespace evio;

// Globals
static std::vector<char*> INFILENAMES;
static char* OUTFILENAME = nullptr;
static bool QUIT = false;

// Prototypes
void ParseCommandLineArguments(int argc, char* argv[]);
void Usage();
void ctrlCHandle(int);
void Process(unsigned int &NEvents, unsigned int &NEvents_read);
std::ifstream::pos_type GetFilesize(const char* filename);

// ---------- GetFilesize ----------
std::ifstream::pos_type GetFilesize(const char* filename) {
    std::ifstream in(filename, std::ifstream::ate | std::ifstream::binary);
    return in.tellg();
}

// ---------- main ----------
int main(int argc, char* argv[]) {
    signal(SIGINT, ctrlCHandle);

    ParseCommandLineArguments(argc, argv);

    // Print input files and their sizes
    for (auto file : INFILENAMES) {
        std::ifstream::pos_type size = GetFilesize(file);
        std::cout << "Input file: " << file << " (size: " << size << " bytes)" << std::endl;
    }

    unsigned int NEvents = 0;
    unsigned int NEvents_read = 0;
    Process(NEvents, NEvents_read);

    return 0;
}

// ---------- ParseCommandLineArguments ----------
void ParseCommandLineArguments(int argc, char* argv[]) {
    INFILENAMES.clear();

    for (int i = 1; i < argc; ++i) {
        char* ptr = argv[i];
        if (ptr[0] == '-') {
            switch (ptr[1]) {
                case 'h': Usage(); break;
                case 'o': OUTFILENAME = &ptr[2]; break;
                default:
                    std::cerr << "Unknown option: " << ptr << std::endl;
                    Usage();
            }
        } else {
            INFILENAMES.push_back(ptr);
        }
    }

    if (INFILENAMES.empty()) {
        std::cerr << "\nYou must specify at least one input file!\n" << std::endl;
        Usage();
    }

    if (!OUTFILENAME) {
        OUTFILENAME = new char[256];
        strcpy(OUTFILENAME, "merged.evio");
    }
}

// ---------- Usage ----------
void Usage() {
    std::cout << "\nUsage:\n";
    std::cout << "  evio_merge_files [-oOutputfile] file1.evio file2.evio ...\n\n";
    std::cout << "Options:\n";
    std::cout << "  -oOutputfile   Set output filename (default: merged.evio)\n";
    std::cout << "\nThis tool merges multiple EVIO files into one output file.\n";
    exit(0);
}

// ---------- ctrlCHandle ----------
void ctrlCHandle(int sig) {
    QUIT = true;
    std::cerr << "\nSIGINT received... exiting soon.\n";
}

// ---------- Process ----------
void Process(unsigned int &NEvents, unsigned int &NEvents_read)
{
    // Basic config
    uint32_t maxRecordBytes     = 1000000;
    uint32_t maxEventsPerRecord = 1000;
    size_t   bufferBytes        = 1000000;
    std::string outFile(OUTFILENAME);

    // Use XML dict from first file, if any
    std::string dictXml = "";
    try {
        if (!INFILENAMES.empty()) {
            EvioReaderV4 dictReader(INFILENAMES[0]);
            if (dictReader.hasDictionaryXML()) {
                dictXml = dictReader.getDictionaryXML();
                std::cout << "Dictionary found in first input file.\n";
            } else {
                std::cout << "No dictionary found in first input file.\n";
            }
        }
    } catch (const std::exception &e) {
        std::cerr << "Error retrieving dictionary from first file: " << e.what() << std::endl;
    }

    // Set up EVIO4 writer
    std::unique_ptr<EventWriterV4> writer = std::make_unique<EventWriterV4>(
        outFile,
        "", "", 1, 0,
        maxRecordBytes, maxEventsPerRecord,
        ByteOrder::ENDIAN_LOCAL,
        dictXml,
        true, false,
        nullptr, 1, 0, 1, 1,
        bufferBytes
    );

    // Loop over all input files
    for (auto filename : INFILENAMES) {
        try {
            std::cout << "Opening input file: " << filename << std::endl;
            EvioReaderV4 reader(filename);
            std::shared_ptr<EvioEvent> event;

            while ((event = reader.parseNextEvent())) {
                writer->writeEvent(event);
                NEvents_read++;
                NEvents++;
            }
        } catch (const std::exception &e) {
            std::cerr << "Error processing file " << filename << ": " << e.what() << std::endl;
            continue;
        }
    }

    writer->close();
    std::cout << "Done. " << NEvents_read << " events read, " << NEvents << " written." << std::endl;
}

File: src/utils/java/evio_merge_files.java
import org.jlab.coda.jevio.*;
import org.jlab.coda.hipo.*;
import java.io.File;
import java.io.IOException;
import java.nio.ByteOrder;

public class evio_merge_files {
    public static void main(String[] args) throws IOException, EvioException {
        // Parse command-line arguments
        if (args.length < 1) {
            System.err.println("You must specify at least one input file!");
            printUsage();
            return;
        }
        String outputFile = "./merged.evio";  // default output name
        java.util.List<String> inputFiles = new java.util.ArrayList<>();
        for (String arg : args) {
            if (arg.startsWith("-h")) {
                printUsage();
                return;
            } 
			else if (arg.startsWith("-o")) {
                // If option is "-oOutputName", set output file name
                outputFile = arg.substring(2);
                System.out.println("Output file: " + outputFile);
                // prepend with "./" if directory not specified
                if (!outputFile.startsWith("./") && !outputFile.startsWith("/")) {
                    // System.out.println("Using current directory");
                    outputFile = "./" + outputFile;
                }
            } 
			else inputFiles.add(arg);
        }
        if (inputFiles.isEmpty()) {
            System.err.println("No input files specified.");
            printUsage();
            return;
        }

        // Print input files and their sizes
        for (String infile : inputFiles) {
            long sizeBytes = new File(infile).length();
            System.out.println("Input file: " + infile + " (size: " + sizeBytes + " bytes)");
        }

        // Retrieve dictionary from the first file (if any)
        String dictionaryXML = "";
        EvioReaderV4 dictReader = new EvioReaderV4(inputFiles.get(0));
        String xml = dictReader.getDictionaryXML();              // get dictionary if present:contentReference[oaicite:24]{index=24}
        if (xml != null) {
            dictionaryXML = xml;
            System.out.println("Dictionary found in first input file.");
        } 
        else System.out.println("No dictionary found in first input file.");
        dictReader.close();  // close dictionary reader

        // Set up EVIO EventWriter for the merged output
        int maxRecordBytes     = 1000000;
        int maxEventsPerRecord = 1000;
        int bufferBytes        = 1000000;
        EventWriterV4 writer = new EventWriterV4(
            outputFile, null, "", 1, 0L,
            maxRecordBytes, maxEventsPerRecord,
            bufferBytes,
            ByteOrder.nativeOrder(), null,
            null, true, false,
            null, 0, 1, 1, 1
        );

        // Process each input file: read all events and write to output
        int eventsRead = 0;
        int eventsWritten = 0;
        for (String filename : inputFiles) {
            EvioReaderV4 reader = new EvioReaderV4(filename);
                System.out.println("Opening input file: " + filename);
                EvioEvent event;
                while ((event = reader.parseNextEvent()) != null) {      // sequentially read events:contentReference[oaicite:25]{index=25}
                    // System.out.printf("Read event %d from %s%n", event.getEventNumber(), filename);
                    writer.writeEvent(event);                            // write event to output:contentReference[oaicite:26]{index=26}
                    eventsRead++;
                    eventsWritten++;
                }
                reader.close();  // close reader for this file
        }

        // Close writer and report
        writer.close();  // finalize output file:contentReference[oaicite:27]{index=27}
        System.out.println("Done. " + eventsRead + " events read, " + eventsWritten + " written.");
    }

    private static void printUsage() {
        System.out.println("\nUsage:");
        System.out.println("  java EvioMergeFiles [-oOutputfile] file1.evio file2.evio ...\n");
        System.out.println("Options:");
        System.out.println("  -oOutputfile   Set output filename (default: merged.evio)\n");
        System.out.println("This tool merges multiple EVIO files into one output file.");
    }
}
